include/oneapi/internal/execution_context_sycl.h

```diff
@@ -41,164 +42,77 @@ namespace interface1
 {
 class OpenClKernelFactory : public Base, public ClKernelFactoryIface
 {
-private:
-    class ProgramCacheEntry
-    {
-    private:
-        OpenClProgramRef * _program;
-
-    public:
-        ProgramCacheEntry() : _program(nullptr) {}
-
-        ~ProgramCacheEntry()
-        {
-            delete _program;
-        }
-
-        void setProgram(OpenClProgramRef *program)
-        {
-            _program = program;
-        }
-
-        OpenClProgramRef * getProgram()
-        {
-            return _program;
-        }
-
-        const char* getName(services::Status * status = nullptr)
-        {
-            if (!_program)
-            {
-                if (status)
-                {
-                    status->add(services::ErrorExecutionContext);
-                }
-                return nullptr;
-            }
-            return _program->getName();
-        }
-    };
-
-    class KernelCacheEntry
-    {
-    private:
-        KernelPtr _kernel;
-        services::String _name;
-
-    public:
-        KernelCacheEntry() : _kernel(), _name() {}
-
-        ~KernelCacheEntry() {}
-
-        void setKernel(KernelPtr kernel, const char *name)
-        {
-            _name = name;
-            _kernel = kernel;
-        }
-
-        KernelPtr getKernel()
-        {
-            return _kernel;
-        }
+public:
+    explicit OpenClKernelFactory(cl::sycl::queue & deviceQueue) : _currentProgramRef(nullptr),
+        _executionTarget(ExecutionTargetIds::unspecified), _deviceQueue(deviceQueue) {}
 
-        const char* getName()
-        {
-            return _name.c_str();
-        }
-    };
+    ~OpenClKernelFactory() DAAL_C11_OVERRIDE {}
 
-public:
-    explicit OpenClKernelFactory(cl::sycl::queue & deviceQueue) : _clProgramRef(nullptr), _executionTarget(ExecutionTargetIds::unspecified), _deviceQueue(deviceQueue)
-    {}
 
-    void build(ExecutionTargetId target, const char * key, const char * program, const char * options = "",
+    void build(ExecutionTargetId target, const char * name, const char * program, const char * options = "",
                services::Status * status = nullptr) DAAL_C11_OVERRIDE
     {
         // TODO: Thread safe?
-        // TODO Rework of "cache"
-
-        uint64_t id = hash(key) % SIZE_CACHE_PROGRAM;
-
-        while (_clProgramCache[id].getProgram() != nullptr && strcmp(_clProgramCache[id].getName(), key) != 0)
-        {
-            id = (id + 1) % SIZE_CACHE_PROGRAM;
-        }
-
-        if (_clProgramCache[id].getProgram())
+        services::Status localStatus;
+        services::String key = name;
+        const bool res = programHashTable.contain(key, localStatus);
+        DAAL_CHECK_LOCAL_STATUS(status, localStatus)
```

Comment:
`DAAL_CHECK_LOCAL_STATUS` seems to be redundant. Recommend to avoid creation of new macros
```cpp
services::Status localStatus;
programHashTable.contain(key, localStatus);
services::internal::tryAssignStatus(status, localStatus);
DAAL_CHECK_STATUS_VAR(localStatus);
```